flow/datatypes/numeric.go

```diff
@@ -1,111 +1,163 @@
+ 6. The scale in this case ranges from -1000 to 1000. Yes, it can be negative. Yes, it can be more than the precision.
 
-func (DefaultNumericCompatibility) MaxPrecision() int16 {
-	return 38
+Currently, no DWH supports the two weird cases of scales in Postgres NUMERICs. Expected is that the 0 <= scale < precision.
+In this case, we will default to the default scale and maximum precision for the DWH.
+*/
+type NumericTypmod struct {
+	constrained bool
+	precision   int16
+	scale       int16
 }
 
-func (DefaultNumericCompatibility) MaxScale() int16 {
-	return 37
-}
-
-func (DefaultNumericCompatibility) DefaultPrecisionAndScale() (int16, int16) {
-	return 38, 20
-}
+// This is to reverse what make_numeric_typmod of Postgres does:
+// logic copied from: https://github.com/postgres/postgres/blob/c4d5cb71d229095a39fda1121a75ee40e6069a2a/src/backend/utils/adt/numeric.c#L929
+// Maps most "invalid" typmods to be unconstrained (same as -1)
+func NewParsedNumericTypmod(typmod int32) *NumericTypmod {
+	if typmod < VARHDRSZ {
+		return &NumericTypmod{
+			constrained: false,
+		}
+	}
 
-func IsValidPrecision(precision int16, warehouseNumeric WarehouseNumericCompatibility) bool {
-	return precision <= warehouseNumeric.MaxPrecision()
+	typmod -= VARHDRSZ
+	// if precision or scale are out of bounds, switch to unconstrained and hope for the best
+	precision := int16((typmod >> 16) & 0xFFFF)
+	scale := int16(((typmod & 0x7ff) ^ 1024) - 1024)
+	if precision < 1 || precision > 1000 || scale < -1000 || scale > 1000 {
+		return &NumericTypmod{
+			constrained: false,
+		}
+	}
+	return &NumericTypmod{
+		constrained: true,
+		precision:   int16((typmod >> 16) & 0xFFFF),
+		scale:       int16(((typmod & 0x7ff) ^ 1024) - 1024),
+	}
 }
 
-func IsValidPrecisionAndScale(precision, scale int16, warehouseNumeric WarehouseNumericCompatibility) bool {
-	return IsValidPrecision(precision, warehouseNumeric) && scale <= warehouseNumeric.MaxScale()
+// responsibility of caller to ensure sensible values are passed in
+func NewConstrainedNumericTypmod(precision int16, scale int16) *NumericTypmod {
+	return &NumericTypmod{
+		constrained: true,
+		precision:   precision,
+		scale:       scale,
+	}
 }
 
-func MakeNumericTypmod(precision int32, scale int32) int32 {
-	if precision == 0 && scale == 0 {
+func (t *NumericTypmod) ToTypmod() int32 {
+	if t == nil || !t.constrained {
 		return -1
 	}
-	return (precision << 16) | (scale & 0x7ff) + VARHDRSZ
+	return ((int32(t.precision) << 16) | (int32(t.scale) & 0x7ff)) + VARHDRSZ
 }
 
-// This is to reverse what make_numeric_typmod of Postgres does:
-// https://github.com/postgres/postgres/blob/21912e3c0262e2cfe64856e028799d6927862563/src/backend/utils/adt/numeric.c#L897
-func ParseNumericTypmod(typmod int32) (int16, int16) {
-	offsetMod := typmod - VARHDRSZ
-	precision := int16((offsetMod >> 16) & 0x7FFF)
-	scale := int16(offsetMod & 0x7FFF)
-	return precision, scale
+func (t *NumericTypmod) PrecisionAndScale() (int16, int16) {
```

Comment:
Can make Precision/Scale public & remove this. No need for nil checks if you usually work with these by value